Sync up of resources in bulk not going through

**Describe the bug**
Issue reported on WDF app on [Sync up of resources in bulk not going through](https://github.com/ona-health/hisp-wdf-fhir-resources/issues/194) ticket.

Executing a  workflow that creates a CarePlan and then updates it within a short time results in sync failure. Both the act of creating and updating the CarePlan result in the creation of  `LocalChangeEntity` entries for the same resource that have the same change log type i.e `INSERT`.
The SDK has a check that prevents the merging of `LocalChangeEntity` records for the same resource that have more than one record of type `INSERT`

The reason that this is not a common bug could be that, in most use-cases, the time between creating and updating  a CarePlan is long enough leading to  the `LocalChangeEntity` records for the 2 processes being processed in different batches. 

**Expected behavior**
Multiple resources registered on the app should sync up successfully.

The following error is logged when manual sync is initiated 
```
Work [ id=06b80031-ae53-4027-b1e7-460b21c86453, tags={ org.smartregister.fhircore.engine.sync.AppSyncWorker } ] failed because it threw an exception/error
                                                                                                    java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Changes before creation of resource are not permitted
                                                                                                    	at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:516)
                                                                                                    	at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:475)
                                                                                                    	at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:317)
                                                                                                    	at androidx.work.impl.utils.SerialExecutorImpl$Task.run(SerialExecutorImpl.java:96)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
                                                                                                    	at java.lang.Thread.run(Thread.java:923)
                                                                                                    Caused by: java.lang.IllegalArgumentException: Changes before creation of resource are not permitted
                                                                                                    	at com.google.android.fhir.sync.upload.patch.PerResourcePatchGenerator.mergeLocalChangesForSingleResource(PerResourcePatchGenerator.kt:50)
                                                                                                    	at com.google.android.fhir.sync.upload.patch.PerResourcePatchGenerator.generate(PerResourcePatchGenerator.kt:39)
                                                                                                    	at com.google.android.fhir.sync.upload.Uploader$upload$2.invokeSuspend(Uploader.kt:49)
                                                                                                    	at com.google.android.fhir.sync.upload.Uploader$upload$2.invoke(Unknown Source:8)
                                                                                                    	at com.google.android.fhir.sync.upload.Uploader$upload$2.invoke(Unknown Source:4)
                                                                                                    	at kotlinx.coroutines.flow.SafeFlow.collectSafely(Builders.kt:61)
                                                                                                    	at kotlinx.coroutines.flow.AbstractFlow.collect(Flow.kt:230)
                                                                                                    	at com.google.android.fhir.sync.FhirSynchronizer$upload$2$1.invokeSuspend(FhirSynchronizer.kt:136)
                                                                                                    	at com.google.android.fhir.sync.FhirSynchronizer$upload$2$1.invoke(Unknown Source:8)
                                                                                                    	at com.google.android.fhir.sync.FhirSynchronizer$upload$2$1.invoke(Unknown Source:4)
                                                                                                    	at kotlinx.coroutines.flow.SafeFlow.collectSafely(Builders.kt:61)
                                                                                                    	at kotlinx.coroutines.flow.AbstractFlow.collect(Flow.kt:230)
                                                                                                    	at com.google.android.fhir.impl.FhirEngineImpl.syncUpload(FhirEngineImpl.kt:144)
                                                                                                    	at com.google.android.fhir.impl.FhirEngineImpl$syncUpload$1.invokeSuspend(Unknown Source:15)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
```

--- a/android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirCarePlanGenerator.kt
+++ b/android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirCarePlanGenerator.kt
@@ -215,9 +215,9 @@ constructor(
         carePlan.contained.clear()
 
         // Save CarePlan only if it has activity, otherwise just save contained/dependent resources
-        if (output.hasActivity()) defaultRepository.create(true, carePlan)
+        if (output.hasActivity()) defaultRepository.addOrUpdate(true, carePlan)
 
-        dependents.forEach { defaultRepository.create(true, it) }
+        dependents.forEach { defaultRepository.addOrUpdate(true, it) }
 
         if (carePlan.status == CarePlan.CarePlanStatus.COMPLETED) {
           carePlan.activity

--- a/android/engine/src/test/java/org/smartregister/fhircore/engine/task/FhirCarePlanGeneratorTest.kt
+++ b/android/engine/src/test/java/org/smartregister/fhircore/engine/task/FhirCarePlanGeneratorTest.kt
@@ -501,8 +501,7 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
 
     val resourcesSlot = mutableListOf<Resource>()
     val booleanSlot = slot<Boolean>()
-    coEvery { defaultRepository.create(capture(booleanSlot), capture(resourcesSlot)) } returns
-      emptyList()
+    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just runs
     coEvery { fhirEngine.get<StructureMap>("131373") } returns structureMap
     coEvery { fhirEngine.search<CarePlan>(Search(ResourceType.CarePlan)) } returns listOf()
 
@@ -578,8 +577,7 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
 
     val resourcesSlot = mutableListOf<Resource>()
     val booleanSlot = slot<Boolean>()
-    coEvery { defaultRepository.create(capture(booleanSlot), capture(resourcesSlot)) } returns
-      emptyList()
+    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just runs
     coEvery { fhirEngine.get<StructureMap>("hh") } returns structureMap
     coEvery { fhirEngine.search<CarePlan>(Search(ResourceType.CarePlan)) } returns listOf()
 
@@ -645,8 +643,7 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
 
     val resourcesSlot = mutableListOf<Resource>()
     val booleanSlot = slot<Boolean>()
-    coEvery { defaultRepository.create(capture(booleanSlot), capture(resourcesSlot)) } returns
-      emptyList()
+    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just runs
     coEvery { fhirEngine.get<StructureMap>("hh") } returns structureMap
     coEvery { fhirEngine.search<CarePlan>(Search(ResourceType.CarePlan)) } returns listOf()
 
@@ -820,8 +817,7 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
     val createdTasksSlot = mutableListOf<Resource>()
     val updatedTasksSlot = mutableListOf<Resource>()
     val booleanSlot = slot<Boolean>()
-    coEvery { defaultRepository.create(capture(booleanSlot), capture(createdTasksSlot)) } returns
-      emptyList()
+    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(createdTasksSlot)) } just runs
     coEvery { defaultRepository.addOrUpdate(any(), capture(updatedTasksSlot)) } just runs
     coEvery { fhirEngine.update(any()) } just runs
     coEvery { fhirEngine.get<StructureMap>("528a8603-2e43-4a2e-a33d-1ec2563ffd3e") } returns
@@ -903,8 +899,7 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
 
     val resourcesSlot = mutableListOf<Resource>()
     val booleanSlot = slot<Boolean>()
-    coEvery { defaultRepository.create(capture(booleanSlot), capture(resourcesSlot)) } returns
-      emptyList()
+    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just runs
     coEvery { fhirEngine.get<StructureMap>("528a8603-2e43-4a2e-a33d-1ec2563ffd3e") } returns
       structureMap
 
@@ -1552,8 +1547,7 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
 
     val resourcesSlot = mutableListOf<Resource>()
     val booleanSlot = slot<Boolean>()
-    coEvery { defaultRepository.create(capture(booleanSlot), capture(resourcesSlot)) } returns
-      emptyList()
+    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just runs
     coEvery { fhirEngine.get<StructureMap>("63752b18-9f0e-48a7-9a21-d3714be6309a") } returns
       structureMap
     coEvery { fhirEngine.search<CarePlan>(Search(ResourceType.CarePlan)) } returns emptyList()
@@ -1584,7 +1578,7 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
         assertEquals(DateTimeType.now().value.makeItReadable(), carePlan.created.makeItReadable())
         assertEquals(patient.generalPractitionerFirstRep.extractId(), carePlan.author.extractId())
         assertTrue(carePlan.activityFirstRep.outcomeReference.isNotEmpty())
-        coEvery { defaultRepository.create(capture(booleanSlot), capture(resourcesSlot)) }
+        coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) }
         resourcesSlot
           .filter { res -> res.resourceType == ResourceType.Task }
           .map { it as Task }
@@ -2091,10 +2085,9 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
       fhirEngine.create(patient)
 
       val resourceSlot = slot<Resource>()
-      coEvery { defaultRepository.create(any(), capture(resourceSlot)) } answers
+      coEvery { defaultRepository.addOrUpdate(any(), capture(resourceSlot)) }  answers
         {
           runBlocking(Dispatchers.IO) { fhirEngine.create(resourceSlot.captured) }
-          listOf()
         }
       val carePlan =
         fhirCarePlanGenerator.generateOrUpdateCarePlan(
@@ -2150,8 +2143,7 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
 
     val resourcesSlot = mutableListOf<Resource>()
     val booleanSlot = slot<Boolean>()
-    coEvery { defaultRepository.create(capture(booleanSlot), capture(resourcesSlot)) } returns
-      emptyList()
+    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just runs
     coEvery { fhirEngine.search<CarePlan>(Search(ResourceType.CarePlan)) } returns listOf()
     coEvery { fhirEngine.get<StructureMap>(structureMapRegister.logicalId) } returns
       structureMapRegister

--- a/android/engine/src/test/java/org/smartregister/fhircore/engine/task/FhirCarePlanGeneratorTest.kt
+++ b/android/engine/src/test/java/org/smartregister/fhircore/engine/task/FhirCarePlanGeneratorTest.kt
@@ -501,7 +501,8 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
 
     val resourcesSlot = mutableListOf<Resource>()
     val booleanSlot = slot<Boolean>()
-    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just runs
+    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just
+      runs
     coEvery { fhirEngine.get<StructureMap>("131373") } returns structureMap
     coEvery { fhirEngine.search<CarePlan>(Search(ResourceType.CarePlan)) } returns listOf()
 
@@ -577,7 +578,8 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
 
     val resourcesSlot = mutableListOf<Resource>()
     val booleanSlot = slot<Boolean>()
-    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just runs
+    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just
+      runs
     coEvery { fhirEngine.get<StructureMap>("hh") } returns structureMap
     coEvery { fhirEngine.search<CarePlan>(Search(ResourceType.CarePlan)) } returns listOf()
 
@@ -643,7 +645,8 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
 
     val resourcesSlot = mutableListOf<Resource>()
     val booleanSlot = slot<Boolean>()
-    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just runs
+    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just
+      runs
     coEvery { fhirEngine.get<StructureMap>("hh") } returns structureMap
     coEvery { fhirEngine.search<CarePlan>(Search(ResourceType.CarePlan)) } returns listOf()
 
@@ -817,7 +820,8 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
     val createdTasksSlot = mutableListOf<Resource>()
     val updatedTasksSlot = mutableListOf<Resource>()
     val booleanSlot = slot<Boolean>()
-    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(createdTasksSlot)) } just runs
+    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(createdTasksSlot)) } just
+      runs
     coEvery { defaultRepository.addOrUpdate(any(), capture(updatedTasksSlot)) } just runs
     coEvery { fhirEngine.update(any()) } just runs
     coEvery { fhirEngine.get<StructureMap>("528a8603-2e43-4a2e-a33d-1ec2563ffd3e") } returns
@@ -899,7 +903,8 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
 
     val resourcesSlot = mutableListOf<Resource>()
     val booleanSlot = slot<Boolean>()
-    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just runs
+    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just
+      runs
     coEvery { fhirEngine.get<StructureMap>("528a8603-2e43-4a2e-a33d-1ec2563ffd3e") } returns
       structureMap
 
@@ -1547,7 +1552,8 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
 
     val resourcesSlot = mutableListOf<Resource>()
     val booleanSlot = slot<Boolean>()
-    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just runs
+    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just
+      runs
     coEvery { fhirEngine.get<StructureMap>("63752b18-9f0e-48a7-9a21-d3714be6309a") } returns
       structureMap
     coEvery { fhirEngine.search<CarePlan>(Search(ResourceType.CarePlan)) } returns emptyList()
@@ -2085,7 +2091,7 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
       fhirEngine.create(patient)
 
       val resourceSlot = slot<Resource>()
-      coEvery { defaultRepository.addOrUpdate(any(), capture(resourceSlot)) }  answers
+      coEvery { defaultRepository.addOrUpdate(any(), capture(resourceSlot)) } answers
         {
           runBlocking(Dispatchers.IO) { fhirEngine.create(resourceSlot.captured) }
         }
@@ -2143,7 +2149,8 @@ class FhirCarePlanGeneratorTest : RobolectricTest() {
 
     val resourcesSlot = mutableListOf<Resource>()
     val booleanSlot = slot<Boolean>()
-    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just runs
+    coEvery { defaultRepository.addOrUpdate(capture(booleanSlot), capture(resourcesSlot)) } just
+      runs
     coEvery { fhirEngine.search<CarePlan>(Search(ResourceType.CarePlan)) } returns listOf()
     coEvery { fhirEngine.get<StructureMap>(structureMapRegister.logicalId) } returns
       structureMapRegister

